rfc/rfc-87/rfc-87.md

```diff
@@ -24,9 +24,329 @@
 ## Approvers
 
 - @danny0405
-- @xiarixiaoyao
-- @yuzhaojing
+- @cshuo
 
 ## Status: Claim
 
-JIRA: [HUDI-8934](https://issues.apache.org/jira/browse/HUDI-8934)
+Umbrella ticket: [HUDI-9075](https://issues.apache.org/jira/browse/HUDI-9075)
+
+## Abstract
+
+Building on RFC-84, which removed Avro from Flink’s pre-write operators, RFC-87 eliminates Avro from the write path to improve performance.
+Current writes suffer from excessive Avro serialization/deserialization and in-memory storage of List<HoodieRecord>, causing high GC overhead.
+This RFC replaces DataBucket’s list storage with Flink’s BinaryInMemorySortBuffer, enabling efficient sorting and iterator-based writes.
+HoodieLogBlock is also refactored to separate deserialization from buffering.
+Precombine field deduplication will now occur after sorting.
```

Comment:
I'm wondering is sorting necessary, currently operations on log like compaction or scanning do not require the data in log file is ordered. Perhaps for now, we only treat `BinaryInMemorySortBuffer` as a buffer, with sorting as a potential capability reserved for future needs.

Reply:
We can keep the sort if the perf does not degrade too much. A sorted parquet is more friendly for query engine optimization and it is pentential that we may introduce LSM-tree like merging in the future.

Reply:
From my benchmarks perf doesn't degrade, it was a kind of free cost sorting. I will retest it and provide benchmark results to came to decision. Perf of sorting depends on average record size. I will provide a result for tpc-h lineitem table.

Reply:
Some additional insights, many users are actually using MOR logs as the substitute for MQs, orders of records  are essential to the business analytic in such cases. It's better to keep non-sorting as the default behavior anyway.

Reply:
Does Hudi guarantee order of records while reading? If keep sorting as default behavior we could achieve merge sort compaction strategy by default. Maybe its better to make sorting as default? @danny0405 @geserdugarov

Reply:
We can sort if the cost is low, but we should keep at least the sequence for same record key records if the ordering value are the same.